fdb-extensions/src/main/java/com/apple/foundationdb/async/RTree.java

```diff
@@ -0,0 +1,2568 @@
+        final List<CompletableFuture<Void>> working = Lists.newArrayList();
+        final int numSiblings = config.getSplitS();
+        final Node[] siblings = new Node[numSiblings];
+
+        //
+        // Do some acrobatics to find the best start/end positions for the siblings. Take into account how many
+        // are warranted, if the node that was passed occupies a slot in its parent node that is touching the end or the
+        // beginning of the parent's slots, and the total number of slots in the parent of the node that was
+        // passed in.
+        //
+        final IntermediateNode parentNode = Objects.requireNonNull(node.getParentNode());
+        final List<ChildSlot> children = parentNode.getSlots();
+        int slotIndexInParent = node.getSlotIndexInParent();
+        int start = slotIndexInParent - numSiblings / 2;
+        int end = start + numSiblings;
+        if (start < 0) {
+            start = 0;
+            end = numSiblings;
+        } else if (end > children.size()) {
+            end = children.size();
+            start = end - numSiblings;
+        }
+
+        // because lambdas
+        final int minSibling = start;
+
+        for (int i = start; i < end; i ++) {
+            toBeProcessed.addLast(children.get(i).getChildId());
+        }
+
+        // Fetch all sibling nodes (in parallel if possible).
+        return AsyncUtil.whileTrue(() -> {
+            working.removeIf(CompletableFuture::isDone);
+
+            while (working.size() <= MAX_CONCURRENT_READS) {
+                final int index = numSiblings - toBeProcessed.size();
+                final byte[] currentId = toBeProcessed.pollFirst();
+                if (currentId == null) {
+                    break;
+                }
+
+                final int slotIndex = minSibling + index;
+                if (slotIndex != slotIndexInParent) {
+                    working.add(fetchNode(transaction, currentId).thenAccept(siblingNode -> {
+                        siblingNode.linkToParent(parentNode, slotIndex);
+                        siblings[index] = siblingNode;
+                    }));
+                } else {
+                    // put node in the list of siblings -- even though node is strictly speaking not a sibling of itself
+                    siblings[index] = node;
+                }
+            }
+
+            if (working.isEmpty()) {
+                return AsyncUtil.READY_FALSE;
+            }
+            return AsyncUtil.whenAny(working).thenApply(v -> true);
+        }, executor).thenApply(vignore -> Lists.newArrayList(siblings));
+    }
+
+    /**
+     * Method to validate the Hilbert R-tree.
+     * @param transaction the transaction to use
+     * @return a completable future that completes successfully if the tree is valid, completes with failure otherwise
+     */
+    @Nonnull
+    public CompletableFuture<Void> validate(@Nonnull final Transaction transaction) {
```

Comment:
At scale, we may need a version of this that spans transactions. This is fine for tests, though.

Reply:
Makes sense. Will take a note of that.